### DISCUSSION TITLE
What is the on_train_dataloader hook expected to do?

### DISCUSSION THREAD

**Discussion by ananthsub** [Lightning-AI/pytorch-lightning#8735] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
https://github.com/PyTorchLightning/pytorch-lightning/blob/963c26764682fa4cf64c93c5a7572ae0040e9c32/pytorch_lightning/core/hooks.py#L669-L670

Why do we have these hooks? What does it mean to “request” these dataloaders? What logic goes here vs inside the train/val/test/predict dataloader methods directly? What guarantees does the trainer make about when or how often these hooks are called?

**Comment by carmocca** [comment]
> Why do we have these hooks?

cc @tchaton who added these in https://github.com/PyTorchLightning/pytorch-lightning/commit/0995d30fab0590d155895a77535663794118b5f6

Seems like they are used to inject logic in Flash

https://github.com/PyTorchLightning/lightning-flash/blob/4ebb0877f3d406919fc07791369f0a33643c1ec5/flash/core/model.py#L504-L526

cc @ethanwharris 

> What does it mean to “request” these dataloaders?

To call `model.*_dataloader()` or `datamodule.*_dataloader()`, the verb request was probably used due to the name of the function that calls it:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4928dc55790eec4ee4a558f494a6fc3d49a5e78a/pytorch_lightning/trainer/data_loading.py#L503

> What guarantees does the trainer make about when or how often these hooks are called?

The only guarantee is that it will get called when the dataloader functions do.

**Reply by ananthsub** [reply]
@ethanwharris @tchaton - since it's called at the same time `model.*_dataloader()` or `datamodule.*_dataloader()` are called, can this logic sit directly in `train_dataloader()`? what justifies the separate hook?

**Reply by ethanwharris** [reply]
@ananthsub I'm not sure on the details, but it seems to me like the `on_*_dataloader` hooks are not needed and could be deprecated. At least, if we deprecate it and it turns out we really need it (or someone else does) then we can always undeprecate haha

**Reply by ananthsub** [reply]
filed https://github.com/PyTorchLightning/pytorch-lightning/issues/8758 to discuss further

**Reply by tchaton** [reply]
Hey @ananthsub, 

Those hooks were used to perform in-place modification every time the data-loaders were to be created.
We could deprecate them as there is other alternative within Flash now. 

Best,
T.C
